keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md

```diff
@@ -0,0 +1,449 @@
+for the custom queue.
+
+2. Provide a new function `WithCustomQueue`, users can register the custom queue with this
+function like other plugins.
+
+    ```go
+         command := app.NewSchedulerCommand(
+             app.WithPlugin(coscheduling.Name, coscheduling.New),
+             ......
+       +     app.WithCustomQueue(customQueue.New),
+         )
+    ```
+
+3. The registered custom queue will be recorded by updated `Registry` struct.
+
+    ```go
+      type Registry struct {
+          Pf          map[string]PluginFactory
+          CustomQueue schedulingQueue.SchedulingQueue
+      }
+    ```
+
+4. In function `(c *Configurator) create()`, custom queue will be passed to the scheduler
+if it exists, or the current internal queue will be used.
+
+    ```go
+      - podQueue := internalqueue.NewSchedulingQueue(
+      -         lessFn,
+      +
+      + var podQueue framework.SchedulingQueue
+      + if c.registry.CustomQueue != nil {
+      +         podQueue = c.registry.CustomQueue
+      + } else {
+      +         podQueue = &internalqueue.PriorityQueue{}
+      + }
+      + podQueue.Init(...)
+    ```
+
+5. Basic inputs for queue's initialization will not be changed, a new method `Init` is added in
+the interface `SchedulingQueue`, users must implement it.
+
+    ```go
+      + func (p *PriorityQueue) Init(
+      +        lessFn framework.LessFunc,
+      +        informerFactory informers.SharedInformerFactory,
+      +        opts ...framework.Option,
+      + ) {
+    ```
+
+### Test Plan
+
+- **Unit Tests**: All core changes must be covered by unit tests.
+- **Integration Tests**: One integration test for the custom queue.
+- **Benchmark Tests**: The performance benchmark test result is same as before if custom queue is not used.
+
+### Graduation Criteria
+
+<!--
+**Note:** *Not required until targeted at a release.*
+
+Define graduation milestones.
+
+These may be defined in terms of API maturity, or as something else. The KEP
+should keep this high-level with a focus on what signals will be looked at to
+determine graduation.
+
+Consider the following in developing the graduation criteria for this enhancement:
+- [Maturity levels (`alpha`, `beta`, `stable`)][maturity-levels]
+- [Deprecation policy][deprecation-policy]
+
+Clearly define what graduation means by either linking to the [API doc
+definition](https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-versioning)
+or by redefining what graduation means.
+
+In general we try to use the same stages (alpha, beta, GA), regardless of how the
+functionality is accessed.
+
+[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
+[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
+
+Below are some examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].
+
+#### Alpha
+
+- Feature implemented behind a feature flag
```

Comment:
I guess we don't need a feature gate here. But we can mention the maturity level by using the terms alpha, beta, etc.

Reply:
Updated as following, is this OK?
```
### Graduation Criteria

#### Alpha -> Beta Graduation

- Users can implement their own custom queues.
- No user complaints regarding correctness.

#### Beta -> GA Graduation

- Allowing time for feedback to ensure that the new interface sufficiently expresses users requirements.
```